File: clients/dotnet/OxiDb.Client.Tcp/Query.cs
namespace OxiDb.Client.Tcp;

/// <summary>
/// Type-safe query builder. Each helper returns a plain
/// <see cref="Dictionary{TKey, TValue}"/> that the wire protocol's
/// query layer accepts directly — anywhere the typed CRUD overloads
/// take an <c>object query</c>, you can pass a <c>Query.*</c> result.
/// </summary>
/// <remarks>
/// The shapes match the JSON query operators documented at
/// <see href="https://oxidb.baltavista.com/docs/">oxidb.baltavista.com/docs</see>:
/// <c>$eq</c>, <c>$ne</c>, <c>$gt</c>, <c>$gte</c>, <c>$lt</c>, <c>$lte</c>,
/// <c>$in</c>, <c>$nin</c>, <c>$exists</c>, <c>$regex</c>, <c>$and</c>,
/// <c>$or</c>, <c>$nor</c>, <c>$not</c>.
///
/// All helpers are pure functions. The returned dictionaries are
/// independent — you can compose them freely.
///
/// <para>
/// Prefer the LINQ provider (<see cref="OxiDb.Linq.OxiDbClientExtensions"/>)
/// when your query naturally maps to LINQ — the source generator picks
/// it up and translates at compile time. Use <see cref="Query"/> when
/// you want a value-typed builder for runtime-constructed queries
/// (e.g. mapping from user input).
/// </para>
/// </remarks>
public static class Query
{
    // ── Equality / comparison ───────────────────────────────────────────

    /// <summary>Match documents where <paramref name="field"/> equals
    /// <paramref name="value"/>. Equivalent to MongoDB's implicit equality.</summary>
    public static Dictionary<string, object?> Eq(string field, object? value)
        => new() { [field] = value };

    /// <summary>Match documents where <paramref name="field"/> is not equal
    /// to <paramref name="value"/>.</summary>
    public static Dictionary<string, object?> Ne(string field, object? value)
        => new() { [field] = new Dictionary<string, object?> { ["$ne"] = value } };

    /// <summary>Match documents where <paramref name="field"/> &gt; <paramref name="value"/>.</summary>
    public static Dictionary<string, object?> Gt(string field, object? value)
        => new() { [field] = new Dictionary<string, object?> { ["$gt"] = value } };

    /// <summary>Match documents where <paramref name="field"/> &gt;= <paramref name="value"/>.</summary>
    public static Dictionary<string, object?> Gte(string field, object? value)
        => new() { [field] = new Dictionary<string, object?> { ["$gte"] = value } };

    /// <summary>Match documents where <paramref name="field"/> &lt; <paramref name="value"/>.</summary>
    public static Dictionary<string, object?> Lt(string field, object? value)
        => new() { [field] = new Dictionary<string, object?> { ["$lt"] = value } };

    /// <summary>Match documents where <paramref name="field"/> &lt;= <paramref name="value"/>.</summary>
    public static Dictionary<string, object?> Lte(string field, object? value)
        => new() { [field] = new Dictionary<string, object?> { ["$lte"] = value } };

    /// <summary>Match documents where <paramref name="field"/>'s value is in
    /// the given set. Index-accelerated.</summary>
    public static Dictionary<string, object?> In<T>(string field, IEnumerable<T> values)
        => new() { [field] = new Dictionary<string, object?> { ["$in"] = values.Cast<object?>().ToArray() } };

    /// <summary>Match documents where <paramref name="field"/>'s value is
    /// NOT in the given set.</summary>
    public static Dictionary<string, object?> Nin<T>(string field, IEnumerable<T> values)
        => new() { [field] = new Dictionary<string, object?> { ["$nin"] = values.Cast<object?>().ToArray() } };

    /// <summary>Match documents where <paramref name="field"/> exists (or
    /// doesn't, when <paramref name="exists"/> is false).</summary>
    public static Dictionary<string, object?> Exists(string field, bool exists = true)
        => new() { [field] = new Dictionary<string, object?> { ["$exists"] = exists } };

    /// <summary>Match documents where <paramref name="field"/>'s string
    /// value matches the given regex pattern.</summary>
    public static Dictionary<string, object?> Regex(string field, string pattern)
        => new() { [field] = new Dictionary<string, object?> { ["$regex"] = pattern } };

    // ── Logical combinators ─────────────────────────────────────────────

    /// <summary>Match documents satisfying ALL sub-queries.</summary>
    public static Dictionary<string, object?> And(params Dictionary<string, object?>[] subQueries)
        => new() { ["$and"] = subQueries };

    /// <summary>Match documents satisfying AT LEAST ONE sub-query.</summary>
    public static Dictionary<string, object?> Or(params Dictionary<string, object?>[] subQueries)
        => new() { ["$or"] = subQueries };

    /// <summary>Match documents satisfying NONE of the sub-queries.</summary>
    public static Dictionary<string, object?> Nor(params Dictionary<string, object?>[] subQueries)
        => new() { ["$nor"] = subQueries };

    // ── Range ───────────────────────────────────────────────────────────

    /// <summary>Match documents where <paramref name="field"/> lies in the
    /// half-open interval <c>[low, high)</c>. Single-field wrapper that
    /// builds <c>{ $gte: low, $lt: high }</c> on the same field.</summary>
    public static Dictionary<string, object?> Range(string field, object? low, object? high)
        => new()
        {
            [field] = new Dictionary<string, object?>
            {
                ["$gte"] = low,
                ["$lt"] = high,
            },
        };
}

File: clients/java/src/main/java/com/oxidb/client/Query.java
package com.oxidb.client;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Type-safe query builder. Each helper returns a {@link Map} that the
 * wire protocol accepts directly — anywhere {@link OxiDbClient}
 * takes a query map, you can pass a {@code Query.*} result.
 *
 * <p>Shapes match the JSON operators documented at
 * <a href="https://oxidb.baltavista.com/docs/">oxidb.baltavista.com/docs</a>:
 * {@code $eq, $ne, $gt, $gte, $lt, $lte, $in, $nin, $exists, $regex,
 * $and, $or, $nor}.</p>
 */
public final class Query {

    private Query() {}

    /** Match docs where {@code field} equals {@code value}. */
    public static Map<String, Object> eq(String field, Object value) {
        return Map.of(field, value);
    }

    /** Match docs where {@code field} is not equal to {@code value}. */
    public static Map<String, Object> ne(String field, Object value) {
        return wrap(field, "$ne", value);
    }

    /** Match docs where {@code field > value}. */
    public static Map<String, Object> gt(String field, Object value) {
        return wrap(field, "$gt", value);
    }

    /** Match docs where {@code field >= value}. */
    public static Map<String, Object> gte(String field, Object value) {
        return wrap(field, "$gte", value);
    }

    /** Match docs where {@code field < value}. */
    public static Map<String, Object> lt(String field, Object value) {
        return wrap(field, "$lt", value);
    }

    /** Match docs where {@code field <= value}. */
    public static Map<String, Object> lte(String field, Object value) {
        return wrap(field, "$lte", value);
    }

    /** Match docs where {@code field}'s value is in the given set. */
    public static Map<String, Object> in(String field, List<?> values) {
        return wrap(field, "$in", values);
    }

    /** Match docs where {@code field}'s value is NOT in the given set. */
    public static Map<String, Object> nin(String field, List<?> values) {
        return wrap(field, "$nin", values);
    }

    /** Match docs where {@code field} exists ({@code exists=false} for the inverse). */
    public static Map<String, Object> exists(String field, boolean exists) {
        return wrap(field, "$exists", exists);
    }

    /** Match docs where {@code field}'s string value matches the regex pattern. */
    public static Map<String, Object> regex(String field, String pattern) {
        return wrap(field, "$regex", pattern);
    }

    /** Match docs satisfying ALL sub-queries. */
    @SafeVarargs
    public static Map<String, Object> and(Map<String, Object>... subQueries) {
        return Map.of("$and", Arrays.asList(subQueries));
    }

    /** Match docs satisfying AT LEAST ONE sub-query. */
    @SafeVarargs
    public static Map<String, Object> or(Map<String, Object>... subQueries) {
        return Map.of("$or", Arrays.asList(subQueries));
    }

    /** Match docs satisfying NONE of the sub-queries. */
    @SafeVarargs
    public static Map<String, Object> nor(Map<String, Object>... subQueries) {
        return Map.of("$nor", Arrays.asList(subQueries));
    }

    /** Half-open interval {@code [low, high)} on a single field. */
    public static Map<String, Object> range(String field, Object low, Object high) {
        Map<String, Object> spec = new LinkedHashMap<>();
        spec.put("$gte", low);
        spec.put("$lt", high);
        return Map.of(field, spec);
    }

    private static Map<String, Object> wrap(String field, String op, Object value) {
        return Map.of(field, Map.of(op, value));
    }
}
